projects/jupyter-server-ydoc/jupyter_server_ydoc/handlers.py

```diff
@@ -459,3 +464,133 @@ async def put(self, path):
         )
         self.set_status(201)
         return self.finish(data)
+
+
+class TimelineHandler(APIHandler):
+    def initialize(
+        self, ystore_class: type[BaseYStore], ywebsocket_server: JupyterWebsocketServer
+    ) -> None:
+        self.ystore_class = ystore_class
+        self.ywebsocket_server = ywebsocket_server
+
+    async def get(self, path: str) -> None:
+        file_id_manager = self.settings["file_id_manager"]
+        file_id = file_id_manager.get_id(path)
+
+        # get format and content_type
+        format = str(self.request.query_arguments.get("format")[0].decode("utf-8"))
+        content_type = str(self.request.query_arguments.get("type")[0].decode("utf-8"))
+        encoded_path = encode_file_path(format, content_type, file_id)
+        try:
+            room_id = room_id_from_encoded_path(encoded_path)
+            room: YRoom = await self.ywebsocket_server.get_room(room_id)
+
+            timestamps = [item[-1] async for item in room.ystore.read()]
+            data = {
+                "roomId": room_id,
+                "timestamps": timestamps,
+            }
+
+            self.set_status(200)
+            self.finish(json.dumps(data))
+
+        except RoomNotFound:
+            return None
+
+
+class TimelineForkHandler(APIHandler):
+    def initialize(self, ywebsocket_server: JupyterWebsocketServer) -> None:
+        self._websocket_server = ywebsocket_server
+
+    async def put(self, room_id):
+        idx = uuid4().hex
+        try:
+            fileType = room_id.split(":")[1]
+
+            mode = str(self.request.query_arguments.get("mode")[0].decode("utf-8"))
+            action = str(self.request.query_arguments.get("action")[0].decode("utf-8"))
+            steps = int(self.request.query_arguments.get("steps")[0].decode("utf-8"))
+
+            if mode == "original":
+                root_room = await self._websocket_server.get_room(room_id)
+
+                updates = [item[0] async for item in root_room.ystore.read()]
+                fork_ydoc = Doc()
+
+                ydoc_factory = YDOCS.get(fileType)
+                if ydoc_factory is not None:
+                    FORK_DOCUMENTS[idx] = ydoc_factory(fork_ydoc)
+
+                undo_manager = FORK_DOCUMENTS[idx].undo_manager
+
+                UNDO_MANAGERS[idx] = undo_manager
```

Comment:
I don't see any entry being deleted from `UNDO_MANAGERS`, which means memory will grow indefinitely. Isn't there a time where an undo manager is not needed anymore, maybe when the user chooses to restore a document to some previous state?